text/0006-wasm-smart-contracts.md

```diff
@@ -0,0 +1,440 @@
+---
+layout: default
+title: Wasm_Smart_Contracts
+parent: RFCs
+---
+
+- Feature Name: Support writing Smart Contracts to be executed within a WebAssembly runtime
+- Start Date: 2020-05-20
+- RFC PR: (leave this empty)
+- Fabric Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC is providing:
+
+* a WebAssembly runtime that can be used as a Chaincode implementation in Fabric
+* implementation of a set of APIs to permit Smart Contracts to be written in Rust and then compiled to Wasm bytecode
```

Comment:
Why specifically Rust? Do we know of significant adoption of Rust within the ecosystem target audience of Hyperledger Fabric developers?

Reply:
I think it has more to do with the wasm ecosystem. Of the high-level, first-class languages, Rust has the most well developed support for wasm. C and C++ have pretty equal support, but Rust has much better javascript interfacing and thus makes it easier to build wasm in Rust. Its a much more manual process when compiling C and C++ to Rust. Outside of C, C++ and Rust there aren't many languages that have full support for wasm yet, making Rust the natural choice at this point.

Reply:
The primary WebAssembly languages these days are Rust, AssemblyScript and C/C++.
The official Go compiler supports WebAssembly but assumes a browser/JavaScript host env.  TinyGo looks like a good option but is still missing support for a few standard packages we might need.

Using Rust to start with is our best bet as most WASM stuff can be found in Rust

Reply:
If you are talking about simplicity to start writing and playing with Wasm, I would suggest AssemblyScript and go-wasmer ( as vm host ). AssemblyScript is very close to TypeScript ( with some limitation related to Wasm ) that compiles to wasm bytecodes.
The toolset is very easy to install and use, the usual nmp install and npm run

We are actually running Fabric +  Go Execution environment ( using  go-wasmer as wasm host ) + AssemblyScript ( wasm guest ).
We wrote type binding between AssemblyScript and the Go host application.

Reply:
Having been down this road (deciding on supporting tech for both host and guest) a couple of times now, Rust certainly makes things easier on both sides. However, I think more important than the compilers is the contract between the guest and the host (see the [waPC](https://github.com/wapc) discussion below).  The ideal route is to design your contract such that the language choice of either the guest or the host can be swapped out without ever breaking the contract.

Reply:
@autodidaddict Absolutely agree